bcipy/display/paradigm/vep/display.py

```diff
@@ -207,44 +268,63 @@ def draw_boxes(self) -> None:
             box.draw()
 
     def stimulate(self) -> List[float]:
-        """Stimulate.
-
-        This is the main display function of the VEP paradigm. It is responsible for drawing the flickering stimuli.
+        """
+        This is the main display function of the VEP paradigm. It is
+        responsible for drawing the flickering stimuli.
 
-        It assumes that the VEP stimuli are already constructed using the _build_vep_corner_stimuli function.
-        These boxes are drawn in the order they are in the list as defined in self.vep."""
+        It assumes that the VEP stimuli are already constructed. These boxes
+        are drawn in the order they are in the list as defined in self.vep.
+        """
         self.trigger_callback.reset()
         self.window.callOnFlip(
             self.trigger_callback.callback,
             self.experiment_clock,
             'VEP_STIMULATE')
         self.static_clock.reset()
-        while self.static_clock.getTime() < self.stimuli_timing[-1]:
+        while self.static_clock.getTime() < self.timing_stimuli:
             for frame in range(self.refresh_rate):
-                for box in self.vep:
-                    if box.code[frame] == 1:
-                        box.frame_on()
-                    else:
-                        box.frame_off()
-
-                # flicker!
-                self.draw_static()
+                for stim in self.vep:
+                    stim.render_frame(frame)
+
+                self.draw_boxes()
+                # self.draw_static()
                 self.window.flip()
+        ended_at = self.static_clock.getTime()
+        # TODO: should we have a trigger for VEP_STIM_END?
```

Comment:
No, we know the stim length

Reply:
We do know the stim length and how long this should take, but the issue that I encountered is that rendering was possibly taking longer than this. I got the timing on my local machine to mostly within the allotted stim length by using the multi-polygon checkerboards, but there is always some variation and I haven't yet tested with the full alphabet. Another trigger here would allow us to confirm that all stim happened with the correct timing, though maybe this should only be used in the timing task.

If we need to make adjustments, there are a couple ways to resolve this:
1. We could render fewer things: omit the static elements; use simpler checkerboard patterns such as 4x4 or 3x3, or use fewer boxes (4 instead of 6).
2. We could do a time check after every frame to ensure that we're still within the stim length. However, this may cut the pattern short and the check adds a little overhead to each frame.